### PR TITLE
enable gocyclo linter

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind cleanup

**What this PR does / why we need it**: Fix code and enable gocyclo linter. Follow up from #518.

**Which issue(s) this PR fixes**:

```
pkg/provider/azure_loadbalancer.go:1231:1: cyclomatic complexity 76 of func `(*Cloud).reconcileLoadBalancer` is high (> 30) (gocyclo)
func (az *Cloud) reconcileLoadBalancer(clusterName string, service *v1.Service, nodes []*v1.Node, wantLb bool) (*network.LoadBalancer, error) {
^
pkg/provider/azure_loadbalancer.go:1857:1: cyclomatic complexity 56 of func `(*Cloud).reconcileSecurityGroup` is high (> 30) (gocyclo)
func (az *Cloud) reconcileSecurityGroup(clusterName string, service *v1.Service, lbIP *string, wantLb bool) (*network.SecurityGroup, error) {
^
pkg/provider/azure.go:372:1: cyclomatic complexity 38 of func `(*Cloud).InitializeCloudFromConfig` is high (> 30) (gocyclo)
func (az *Cloud) InitializeCloudFromConfig(config *Config, fromSecret bool) error {
^
pkg/provider/azure_loadbalancer.go:1623:1: cyclomatic complexity 34 of func `(*Cloud).checkLoadBalancerResourcesConflicted` is high (> 30) (gocyclo)
func (az *Cloud) checkLoadBalancerResourcesConflicted(
^
pkg/provider/azure_test.go:2015:1: cyclomatic complexity 33 of func `validateConfig` is high (> 30) (gocyclo)
func validateConfig(t *testing.T, config string) {
^
pkg/provider/azure_loadbalancer.go:2306:1: cyclomatic complexity 32 of func `(*Cloud).reconcilePublicIP` is high (> 30) (gocyclo)
func (az *Cloud) reconcilePublicIP(clusterName string, service *v1.Service, lbName string, wantLb bool) (*network.PublicIPAddress, error) {
^
pkg/provider/azure_test.go:2890:1: cyclomatic complexity 31 of func `TestCanCombineSharedAndPrivateRulesInSameGroup` is high (> 30) (gocyclo)
func TestCanCombineSharedAndPrivateRulesInSameGroup(t *testing.T) {
^
```

Fixes #

**Special notes for your reviewer**:

**Release note**:
```
NONE
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
this is rebased on top of #522 to minimize merge conflicts.

/hold until #522 merges so I can drop the first commit.

File: pkg/provider/azure_loadbalancer.go [hunk 1]
```diff
@@ -1242,98 +1242,217 @@ func (az *Cloud) reconcileLoadBalancer(clusterName string, service *v1.Service,
 	}
 	lbName := *lb.Name
 	lbResourceGroup := az.getLoadBalancerResourceGroup()
+	lbBackendPoolID := az.getBackendPoolID(lbName, az.getLoadBalancerResourceGroup(), getBackendPoolName(clusterName, service))
```
Comment (CecileRobertMichon) [inline pkg/provider/azure_loadbalancer.go:1245]: I had to do some pretty heavy refactoring of the `reconcileLoadBalancer` function (cyclo complexity was 76 before!) to get it under 30, please review extra closely.

**Review by feiskyer — COMMENTED** [review]
This is awesome! Thanks for the refactoring.

/lgtm
/approve

**Pull request merged**